FWCore/Framework/src/LuminosityBlockProcessingStatus.cc

```diff
@@ -23,7 +21,10 @@ namespace edm {
       iter.reset();
     }
     resumeGlobalLumiQueue();
-    run_.reset();
+  }
+
+  void LuminosityBlockProcessingStatus::setGlobalEndRunHolder() {
```

Comment:
Seems strange that we have a `set` function that only gets its data internally.

Reply:
From looking at the code, why does getting the waiting task have to be deferred?

Reply:
After staring at this for a while, I think the simplest thing to do is remove the runStatus_ member variable from LuminosityBlockProcessingStatus. I can propagate the run status object to the places where it is needed without much difficulty. There are not too many places that use it and it is probably simpler and easier to understand than the current version. I will do that now.

The purpose of that WaitingTaskHolder is force global end run to wait for all contained global end lumi transitions to complete. I defer setting the the WaitingTaskHolder until that point in time because that is point where we know the global end lumi transition is actually going to occur. If an exception occurs before that, then we just bail out for that lumi and never call global end lumi (or anything else). We probably could set it earlier, but then we would be waiting for no reason on the status destructor when there was an exception. Probably it would work either way...